stonesoup/sensormanager/base.py

```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+
+from abc import abstractmethod, ABC
+from typing import Callable, Set
+import random
+import numpy as np
+import itertools as it
+
+from ..base import Base, Property
+from ..sensor.sensor import Sensor
+
+
+class SensorManager(Base, ABC):
+    """The sensor manager base class.
+
+    The purpose of a sensor manager is to return a mapping of sensors and sensor actions
+    appropriate to a specific
+    scenario and with a particular objective, or objectives, in mind. This involves using
+    estimates of the situation and knowledge of the sensor system to calculate metrics associated
+    with actions, and then determine optimal, or near optimal, actions to take.
+
+    There is considerable freedom in both the theory and practice of sensor management and these
+    classes do not enforce a particular solution. A sensor manager may be 'centralised' in that
+    it controls the actions of multiple sensors, or individual sensors may have their own managers
+    which communicate with other sensor managers in a networked fashion.
+
+    """
+    sensors: Set[Sensor] = Property(doc="The sensor(s) which the sensor manager is managing. "
+                                        "These must be capable of returning available actions.")
+
+    reward_function: Callable = Property(
+        default=None, doc="A function or class designed to work out the reward associated with an "
+                          "action or set of actions. This may also incorporate a notion of the "
+                          "cost of making a measurement. The values returned may be scalar or "
+                          "vector in the case of multi-objective optimisation. Metrics may be of "
+                          "any type and in any units.")
+
+    @abstractmethod
+    def choose_actions(self, *args, **kwargs):
```

Comment:
I think its worth defining the inputs for this method to try and standardise it.

Reply:
Good point - the RandomSensorManager takes in tracks_list even though it's not required, do we want to keep it that way and have the standard format as (self, tracks_list, timestamp, nchoose=1, *args, **kwargs) ? Alternatively, timestamp is the only input that I can think would always be required so I can just add that in

Reply:
I think we should try to standardise it. No issues with RandomSensorManager just ignoring the variables it doesn't need.

Reply:
Qs: what would you standardise at? Isn't `def choose_actions(self, *args, **kwargs):` standard (albeit flexible)?

I'd prefer not to pass variables that aren't used (yes, I may be responsible for that happening with the RandomSensorManager). It may not have an impact computationally, but it could end up being thoroughly confusing.

Reply:
So we can leave space to expand, but having at least a base set of inputs (so like in the above: tracks and timestamps, and number of choices) then we can create a number of components that will be broadly compatible and interchangeable.

Reply:
We could have timestamps and no. choices as base set of inputs and remove tracks_list (also remove from random one)?

Reply:
timestamp makes sense - you'd want to know when the choice is applicable. nchoose=1 would also seem an appropriate default.

Reply:
I think `tracks` should be in the base set of inputs, and its dynamic information that represents your current understanding (noting that random sensor manager is really a special case, as making random choices is arguably not managing anything :wink: )

Reply:
One goal of the framework is to enable ability to have a set of common components that can be easily interchanged. Whilst excepting that all possible combinations and use cases can't be covered, I think that have tracks as part of that initial baseline is reasonable (and otherwise the existing example we have is itself an exception).

Reply:
> One goal of the framework is to enable ability to have a set of common components that can be easily interchanged. Whilst excepting that all possible combinations and use cases can't be covered, I think that have tracks as part of that initial baseline is reasonable (and otherwise the existing example we have is itself an exception).

but we do have the random sensor manager in the existing example.

Maybe my understanding isn't correct but you're advocating:
`def choose_actions(self, tracks, timestamp, nchoose, *args, **kwargs):`
over 
`def choose_actions(self, *args, **kwargs)`. Is that right? (or close?)

I'd argue that the first example does not promote easy interchange for those (many) sensor managers that are not track-focussed. (Accepting that the difference is only passing an empty set vs passing an extra keyword, I think the preference should be for general, simple, base classes.) But maybe I'm missing something. Is there an obvious problem with a user-defined `choose_actions(timestamp, tracks=some_tracks)` (say)?

Reply:
Yes. My main issue is with `*args` as positional arguments really need to be fixed (as position dependent). Maybe having `choose_actions(timestamp, nchoose, **kwargs)`? That works for your example, just need to be sure that keyword arguments are passed.

Reply:
So put timestamp and nchoose into the base class function, and remove *args from all choose_actions() ?

Reply:
> So put timestamp and nchoose into the base class function, and remove *args from all choose_actions() ?

Happy with that (nchoose=1 as default?). 

Realised I didn't really know what *args does.